cob_twist_controller/src/inverse_jacobian_calculations/inverse_jacobian_calculation.cpp

```diff
@@ -61,47 +63,88 @@ Eigen::MatrixXd PInvBySVD::calculate(const Eigen::MatrixXd& jacobian) const
+        case(SA_NF):
         {
-            // pow(beta, 2) << pow(lambda, 2)
-            singularValuesInv(i) = singularValues(i) / (pow(singularValues(i), 2) + pow(params.beta, 2));
+            // Formula 20 Singularity-robust Task-priority Redundandancy Resolution
+            // Sum part
+            for (uint32_t i = 0; i < singularValues.rows()-1; ++i)
+            {
+                // pow(beta, 2) << pow(lambda, 2)
+                singularValuesInv(i) = singularValues(i) / (pow(singularValues(i), 2) + pow(params.beta, 2));
+            }
+            // Formula 20 - additional part - numerical filtering for least singular value m
+            uint32_t m = singularValues.rows()-1;
+            singularValuesInv(m) = singularValues(m) / (pow(singularValues(m), 2) + pow(params.beta, 2) + pow(lambda, 2));
+
+            result = svd.matrixV() * singularValuesInv.asDiagonal() * svd.matrixU().transpose();
+            break;
         }
-        // Formula 20 - additional part - numerical filtering for least singular value m
-        uint32_t m = singularValues.rows();
-        singularValuesInv(m) = singularValues(m) / (pow(singularValues(m), 2) + pow(params.beta, 2) + pow(lambda, 2));
-    }
-    else
-    {
-        // small change to ref: here quadratic damping due to Control of Redundant Robot Manipulators : R.V. Patel, 2005, Springer [Page 13-14]
-        for (uint32_t i = 0; i < singularValues.rows(); ++i)
+        case(SA_SR):
         {
-            double denominator = (singularValues(i) * singularValues(i) + pow(lambda, 2) );
-            // singularValuesInv(i) = (denominator < eps_truncation) ? 0.0 : singularValues(i) / denominator;
-            singularValuesInv(i) = (singularValues(i) < eps_truncation) ? 0.0 : singularValues(i) / denominator;
+            Eigen::VectorXd lambda_vec = Eigen::VectorXd::Zero(singularValues.size());
+
+            Eigen::MatrixXd S = Eigen::MatrixXd::Zero(singularValues.size(),singularValues.size());
+            for(unsigned i = 0; i < singularValues.size(); i++)
+            {
+                distance = abs(abs(static_cast <double>(joint_states.current_q_(i))) - static_cast <double>(limiter_params.limits_max[i]));
+//                alpha = 1/(1 + exp(((distance - params.damping_delta )/ params.damping_slope)*(-1)));
+                alpha = 1/(1 + exp(((distance - 0.1)/ 0.01)*(-1)));
+
+                ROS_WARN_STREAM("alpha[" << i << "]: " << alpha << "   distance[" << i << "]: " << distance);
+
+                lambda_vec(i) = params.damping_gain /( 1+ exp(static_cast <double>(singularValues(i)) + params.damping_delta) / params.damping_slope);
+                S(i,i) = singularValues(i)/(pow(static_cast <double>(singularValues(i)),2) + lambda_vec(i));
+                v_ith_row = svd_V.row(i) * alpha;
+                svd_V.row(i) = v_ith_row;
+            }
+            result = svd_V * S * svd.matrixU().transpose();
+            break;
+        }
```

Comment:
@ipa-bfb 
Please provide link to literature! Where do these formulae come from?